gateway/mock_response.go

```diff
@@ -0,0 +1,50 @@
+package gateway
+
+import (
+	"bytes"
+	"io"
+	"net/http"
+
+	"github.com/TykTechnologies/tyk/apidef/oas"
+	header "github.com/TykTechnologies/tyk/headers"
+	"github.com/getkin/kin-openapi/routers"
+)
+
+func (p *ReverseProxy) mockResponse(r *http.Request, route *routers.Route) (res *http.Response, err error) {
```

Comment:
Why it is implemented as ReverseProxy code? Imho it should be middleware which sits in the end of the chain. See how it works in old mock plugin, you can write to response inside middleware, and then respond with `mwStatusRespond` which will stop MW chain and it will not reach reverse proxy.

```
    v.DoMockReply(w, meta)
    return nil, mwStatusRespond
```

Reply:
The idea here was not like to create it as a middleware. When you put it as middleware and return the response chains don't work. This way you can manipulate even response, add global headers etc.

Reply:
I see, hm. Have you considered then fix response middleware flow, and make it run when we respond with `mwStatusRespond` ?

Reply:
To be frank looking at `mwStatusRespond` current implementation, it is really bad, in a sense that it litelarly skips everything,probably including analytics. E.g. most likely you won't see legacy mock requests in log browser, and bunch of other stuff. And actually passing it though reverse proxy does makes sense. 
But I wonder if we still can make it work as middleware, but if `mwStatusRespond` is set, it will actually call reverse proxy, it will set some flag to request context or similar, which will automatically skip all middlewares except reverse proxy handler  (can be added inside `createMiddleware`, which will not require modifying all the Middlewares manually). And then this section:
```
			// Special code, bypasses all other execution
			if errCode != mwStatusRespond {
				// No error, carry on...
				meta["bypass"] = "1"
				h.ServeHTTP(w, r)
			} else {
				mw.Base().UpdateRequestSession(r)
			}
```
Will turn into smth like:
```
			// Special code, bypasses all other execution
			if errCode == mwStatusRespond {
                             setCtxValue(req, Response, ResponseProxyTypeObject)
                        }

                        // No error, carry on...
                        meta["bypass"] = "1"
                        h.ServeHTTP(w, r)
```

And after inside `SuccessHandler.ServeHTTP` you will have smth like:
``` 
        if val, found := getCtxValue(req, Response); found {
          resp = val
        } else {
	  resp := s.Proxy.ServeHTTP(w, r)
        }
```
Or maybe easier add it to `ReverseProxy.WrappedServeHTTP`, which skips all the HTTP call stuff, but it use response object from mock/mw.

But obviously it can look more complex as it looks...

Reply:
What is the harm of the current code ? Mocking the outbound request?